### PR TITLE
Exit after run to prevent hanging if there are active listeners

### PR DESCRIPTION
**Pull request by dhadka**
Closes https://github.com/actions/cache/issues/370

If there are active listeners after `run()` finishes, NodeJS will not exit.  Generally, leaving an active listener is a bug that should be fixed, but to prevent this from causing the step to hang we can explicitly exit.

### CODE REVIEW

File: src/restore.ts [hunk 1]
```diff
@@ -60,6 +60,6 @@ async function run(): Promise<void> {
     }
 }
 
-run();
+run().then(() => process.exit(0));
```
Comment (joshmgross) [inline src/restore.ts:63]: I think this is the wrong approach here. If we're confident that these hanging events are only from error cases, then we should have a corresponding `process.exit` in the `catch` blocks. For save, we'd want a success status code and for restore we'd want a failure status code.

The downside of this current implementation is that we're exiting with a success status code when the step may have failed.

**Comment by joshmgross** [comment]
Could you double-check the generated dist files? I wouldn't expect 100k lines to be changed for those.

**Comment by jclem** [comment]
Whoops, this should be `process.exit()` so that we honor any `process.exitCode` previously set.

File: src/restore.ts (hunk 1 above)
Comment (joshmgross) [reply src/restore.ts:63]: Sounds like `process.exit()` is what we want to honor the exit code set by `core.setFailed`
https://github.com/actions/cache/pull/371#issuecomment-658296834

**Comment by jclem** [comment]
Ah, as @ericsciple pointed out, we need to flush stdout/stderr, so just calling `process.exit()` isn't safe.

**Comment by jclem** [comment]
Another option would be to set a timeout for writing stdout/stderr and then force exit:

```typescript
const OUTPUT_TIMEOUT = 60_000

function runSafeMain(fn: (...args: any[]) => Promise<unknown>) {
  fn().finally(waitForOutput)
}

function waitForOutput() {
  const timeout = setTimeout(() => {
    console.error('Output timeout exceeded')
    process.exit()
  }, OUTPUT_TIMEOUT)

  timeout.unref()
}
```

**Comment by dhadka** [comment]
@joshmgross Fixed the `dist/` files having tons of changes. Had it pointing to a local copy of `@actions/toolkit`.  (The test that looks for uncommitted changes also detected this issue).

I'll wait until I hear back from `@jclem` and `@ericsciple` on how to handle flushing before doing anything with this.

**Comment by aiqiaoy** [comment]
```
[warning]downloadCache failed: Request timeout: /758370072ebc48a1bae6dd969702511b/59eed5cca5c5ea119b0600155dafe08a?sv=2018-03-28&sr=b&sig=4O4ae0FEDWtyXHalSOBL7FMQNrPJdJSajOxwzi2WJsY%3D&se=2020-07-14T11%3A18%3A50Z&sp=r&rscl=x-e2eid-ca91503f-2fdf483a-86d18754-8bf30f0c
8
##[error]The operation was canceled.
```

It seems from the log the request got stuck after the timeout error. I suspect it's because of a bug in the http-client package we use (code [here](https://github.com/actions/http-client/blob/5f61f72a73224ee5541c6961f864e09e9f4cf6d7/index.ts#L498)) but I'm not proficient enough in nodejs runtime to figure out where.. @jclem Any idea?

**Comment by aiqiaoy** [comment]
@dhadka I'm changing default branch to `main` for cache so I rebased your PR to on top of `main`

**Comment by bishal-pdMSFT** [comment]
@vsvipul are you reactivating this PR?

**Comment by vsvipul** [comment]
@bishal-pdMSFT  I'm not sure if this is still required. I'll check and follow up accordingly.

**Comment by vsvipul** [comment]
Closing this for now, as we aren't facing any issues related to this.

**Pull request closed without merging**